quantization/src/encoded_vectors_pq.rs

```diff
@@ -97,80 +98,102 @@ impl<TStorage: EncodedStorage> EncodedVectorsPQ<TStorage> {
-                storage_builder.push_vector_data(&encoded);
-                encoded_pool.push(encoded);
-                vectors_pool.push(vector);
-                end_index = next(end_index);
-                busy_threads_count -= 1;
-            }
+    }
 
-            let encoded = encoded_pool.pop().unwrap_or_default();
-            let mut v = vectors_pool
-                .pop()
-                .unwrap_or_else(|| vec![0.0; vector_parameters.dim]);
-            v.copy_from_slice(vector_data);
-            threads[start_index]
-                .vector_sender
-                .send((v, encoded))
-                .unwrap();
-            start_index = next(start_index);
-            busy_threads_count += 1;
-        }
-        for _ in 0..busy_threads_count {
-            let (_, encoded) = threads[end_index].encoded_receiver.recv().unwrap();
-            storage_builder.push_vector_data(&encoded);
-            end_index = next(end_index);
+    /// Encode whole storage inside rayon context
+    /// This function should be called inside `rayon::scope`
+    fn encode_storage_rayon<'a: 'b, 'b>(
+        scope: &rayon::Scope<'b>,
+        data: impl Iterator<Item = &'a [f32]> + Clone + Send + 'b,
+        storage_builder: &'b mut (impl EncodedStorageBuilder<TStorage> + Send),
+        vector_division: &'b [Range<usize>],
+        centroids: &'b [Vec<f32>],
+        max_threads: usize,
+    ) -> Result<(), EncodingError> {
+        let storage_builder = Arc::new(Mutex::new(storage_builder));
+
+        // Synchronization between threads. Use conditional variable for
+        // each thread. Each condvar is blocked instead of first.
+        // While encoding, thread `N` after `storage_builder` usage blocks themself and 
+        // unblock thread `N+1`.
+        // In summary, access to `storage_builder` is ordered by `thread_index` below.
+        let condvars = (0..max_threads)
+            .map(|_| Arc::new(ConditionalVariable::default()))
+            .collect::<Vec<_>>();
+        condvars[0].notify(); // Allow first thread to use storage
+
+        for thread_index in 0..max_threads {
+            // Thread process vectors `N` that `(N + thread_index) % max_threads == 0`.
+            let data = data.clone().skip(thread_index);
+            let storage_builder = storage_builder.clone();
+            let condvar = condvars[thread_index].clone();
+            let next_condvar = condvars[(thread_index + 1) % max_threads].clone();
+
+            scope.spawn(move |_| {
+                let mut encoded_vector = Vec::new();
+                for vector in data.step_by(max_threads) {
+                    encoded_vector.clear();
+                    Self::encode_vector(vector, vector_division, centroids, &mut encoded_vector);
+                    // wait for permission from prev thread to use storage
+                    condvar.wait();
+                    // push encoded vector to storage
+                    storage_builder
```

Comment:
Since we're guaranteed to have exclusive access here, we can probably remove locking.

[`UnsafeCell`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html) may be a good fit here, rather than the `Mutex`.